src/hypothesis/extra/datetime.py

```diff
@@ -15,111 +15,412 @@
+            min_datetime = dt.datetime.min
+        else:
+            naive = dt.datetime.min + dt.timedelta(days=1)
+            min_datetime = min([tz.normalize(tz.localize(naive))
+                                for tz in timezones])
+            min_datetime -= dt.timedelta(days=1)
+    else:
+        assert min_datetime is None
+        assert max_datetime is None
+        min_datetime = dt.datetime.min
+        max_datetime = dt.datetime.max
+
+    # Check that our bounds are of the correct type and ordered correctly
+    check_valid_interval(min_datetime, max_datetime,
+                         'min_datetime', 'max_datetime')
+
+    # Check that allow_naive and timezones args will work with aware bounds
+    if min_datetime.tzinfo is not None:
+        if allow_naive:
+            raise InvalidArgument(
+                'Cannot create naive datetimes between tz-aware bounds as '
+                'they are not comparable; do not set allow_naive=True and '
+                'allow at least one timezone.')
+
+    # The final step of validation is to ensure that every drawable datetime
+    # can be represented in at least one of the provided timezones.
+    # This is an expensive check, but it's check now or error later.
+    if min_datetime.tzinfo is not None:
+        msg = ('It is impossible to localise %s=%s in any of the given '
+               'timezones, illegally reducing the range of values which may '
+               'otherwise be drawn.')
+        if not any(is_representable(min_datetime, tz) for tz in timezones):
+            raise InvalidArgument(msg % ('min_datetime', min_datetime))
+        if not any(is_representable(max_datetime, tz) for tz in timezones):
+            raise InvalidArgument(msg % ('max_datetime', max_datetime))
+
+    # If there is only a single possible date, `just` is more efficient
+    if allow_naive and not timezones:
+        if min_datetime == max_datetime:
+            return just(min_datetime)
+    elif not allow_naive and len(timezones) == 1:
+        to_tz = timezones[0]
+        return just(to_tz.normalize(min_datetime.astimezone(to_tz)))
+
     return DatetimeStrategy(
         allow_naive=allow_naive, timezones=timezones,
-        min_year=min_year, max_year=max_year,
+        min_datetime=min_datetime, max_datetime=max_datetime,
     )
 
 
 @defines_strategy
-def dates(min_year=None, max_year=None):
-    """Return a strategy for generating dates."""
-    return datetimes(
-        allow_naive=True, timezones=[],
-        min_year=min_year, max_year=max_year,
-    ).map(datetime_to_date)
+def dates(min_year=None, max_year=None, min_date=None, max_date=None):
+    """Return a strategy for generating dates.
 
+    The arguments are as for :py:func:`datetimes`, including the deprecation
+    of ``min_year`` and ``max_year``.  If ``min_date`` or ``max_date`` are
+    datetime objects, the time and any timezone information is discarded and
+    they are treated as dates.
```

Comment:
Arguments don’t match `datetimes`, because there aren’t `allow_naive` or `timezones` arguments.

I’d drop the second sentence, I don’t think it adds anything, and it’s hard to imagine something else we might do in such circumstances.